src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,410 @@
+            //
+            // where inner sequence is:
+            //      var leftValue = left.GetValueOrDefault(); (or left if !leftNullable)
+            //      var rightValue = right.GetValueOrDefault(); (or right if !rightNullable)
+            //      ... logical expression using leftValue and rightValue ...
+            //
+            // and true/false and false/true depend on operatorKind (== vs. !=)
+            //
+            // and temps are created an initialized with early side-effects for:
+            //  - elements from left and right, in case of a tuple literal
+            //  - left and right, in case of a tuple or nullable tuple type
+            //
+            // But if neither is nullable, then just produce the inner sequence.
+            //
+            // Note: all the temps are created in a single bucket (rather than different scopes of applicability) for simplicity
+
+            var outerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+            var innerEffects = ArrayBuilder<BoundExpression>.GetInstance();
+
+            BoundExpression leftHasValue;
+            BoundExpression leftValue;
+
+            var coreLeft = WithoutImplicitNullableConversions(left);
+            var isLeftNullable = coreLeft.Type.IsNullableType();
+            if (isLeftNullable)
+            {
+                BoundExpression savedLeft = EvaluateSideEffectingArgumentToTemp(VisitExpression(left), outerEffects, ref initialEffectsAndTemps.temps);
+                leftHasValue = MakeHasValueTemp(savedLeft, initialEffectsAndTemps.temps, outerEffects);
+                leftValue = MakeValueOrDefaultTemp(savedLeft, initialEffectsAndTemps.temps, innerEffects);
+                inLeftLiteral = false;
+            }
+            else
+            {
+                leftHasValue = MakeBooleanConstant(left.Syntax, true);
+                leftValue = SaveTupleIfNotLiteral(coreLeft, ref inLeftLiteral, initialEffectsAndTemps, isRight: false);
+            }
+
+            BoundExpression rightHasValue;
+            BoundExpression rightValue;
+
+            var coreRight = WithoutImplicitNullableConversions(right);
+            var isRightNullable = coreRight.Type.IsNullableType();
+            if (isRightNullable)
+            {
+                BoundExpression savedRight = EvaluateSideEffectingArgumentToTemp(VisitExpression(right), outerEffects, ref initialEffectsAndTemps.temps);
+                rightHasValue = MakeNullableHasValue(savedRight.Syntax, savedRight); // no need for local for right.HasValue since used once
+                rightValue = MakeValueOrDefaultTemp(savedRight, initialEffectsAndTemps.temps, innerEffects);
+                inRightLiteral = false;
+            }
+            else
+            {
+                rightHasValue = MakeBooleanConstant(right.Syntax, true);
+                rightValue = SaveTupleIfNotLiteral(coreRight, ref inRightLiteral, initialEffectsAndTemps, isRight: true);
+            }
```

Comment:
There are several small methods that are only used in the `isLeftNullable` and `isRightNullable` cases above (`MakeHasValueTemp`, `MakeNullableHasValue`, `MakeValueOrDefaultTemp`, `SaveTupleIfNotLiteral`). Would the code be slightly simpler if the two cases were two calls to a single helper method? (It looks like the two cases differ only by whether the `MakeNullableHasValue` result is stored in a temp.) Then the small methods would have one caller each and could be inlined. Just a suggestion.

Reply:
I've removed `isRight` parameters.
As for extracting a method, we'd end up with 9 parameters (some by-val and some out). I didn't feel it improved the code :-(

---
In reply to: [170411915](https://github.com/dotnet/roslyn/pull/23356#discussion_r170411915) [](ancestors = 170411915)